Integers are overflowed silently

Right now this library silently overflows integers. For example, filling in `300` for an `uint8` results in `44`, since the maximum size is `256` and `300 - 256 = 44`.

Personally I think this behaviour is rather surprising, and would expect an error from this library.

Thoughts?

Here's a simple test case:

```
diff --git i/formam_test.go w/formam_test.go
index 470acfa..f2657c9 100644
--- i/formam_test.go
+++ w/formam_test.go
@@ -870,3 +870,22 @@ func TestPanic(t *testing.T) {
        }

 }
+
+// #25
+func TestOverflow(t *testing.T) {
+       s := struct {
+               N uint8
+       }{}
+
+       vals := url.Values{
+               "N": []string{"300"},
+       }
+
+       dec := formam.NewDecoder(&formam.DecoderOptions{})
+       err := dec.Decode(vals, &s)
+       if err != nil {
+               t.Error(err)
+       }
+
+       t.Fatalf("s.N is %d", s.N) // 44
+}
```

Output:

```
[~/code/formam](master)% go test
--- FAIL: TestOverflow (0.00s)
    formam_test.go:890: s.N is 44
FAIL
```

At first I thought the user could detect this, so I created an issue to discus possible options first, but then I realized it's not so easy and that returning an error is really the only sane thing to do (as far as I can tell?)

I'll take a look at fixing it later, after the errcode PR is done (otherwise it'll conflict, which means more work).

--- a/errors.go
+++ b/errors.go
@@ -13,6 +13,7 @@ const (
 	ErrCodeConversion                // Error converting field to the type.
 	ErrCodeUnknownType               // Unknown type.
 	ErrCodeUnknownField              // No struct field for passed parameter (will never be used if IgnoreUnknownKeys is set).
+	ErrCodeRange                     // Number is out of range (e.g. parsing 300 in uint8 would overflow).
 )
 
 type Error struct {

--- a/formam.go
+++ b/formam.go
@@ -418,20 +418,32 @@ func (dec *Decoder) decode() error {
 	case reflect.String:
 		dec.curr.SetString(dec.values[0])
 	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
-		if num, err := strconv.ParseInt(dec.values[0], 10, 64); err != nil {
-			return newError(ErrCodeConversion, dec.field, dec.path, "could not parse number: %s", err)
+		if num, err := strconv.ParseInt(dec.values[0], 10, dec.curr.Type().Bits()); err != nil {
+			code := ErrCodeConversion
+			if err, ok := err.(*strconv.NumError); ok && err.Err == strconv.ErrRange {
+				code = ErrCodeRange
+			}
+			return newError(code, dec.field, dec.path, "could not parse number: %s", err)
 		} else {
 			dec.curr.SetInt(num)
 		}
 	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
-		if num, err := strconv.ParseUint(dec.values[0], 10, 64); err != nil {
-			return newError(ErrCodeConversion, dec.field, dec.path, "could not parse number: %s", err)
+		if num, err := strconv.ParseUint(dec.values[0], 10, dec.curr.Type().Bits()); err != nil {
+			code := ErrCodeConversion
+			if err, ok := err.(*strconv.NumError); ok && err.Err == strconv.ErrRange {
+				code = ErrCodeRange
+			}
+			return newError(code, dec.field, dec.path, "could not parse number: %s", err)
 		} else {
 			dec.curr.SetUint(num)
 		}
 	case reflect.Float32, reflect.Float64:
 		if num, err := strconv.ParseFloat(dec.values[0], dec.curr.Type().Bits()); err != nil {
-			return newError(ErrCodeConversion, dec.field, dec.path, "could not parse float: %s", err)
+			code := ErrCodeConversion
+			if err, ok := err.(*strconv.NumError); ok && err.Err == strconv.ErrRange {
+				code = ErrCodeRange
+			}
+			return newError(code, dec.field, dec.path, "could not parse float: %s", err)
 		} else {
 			dec.curr.SetFloat(num)
 		}

--- a/formam_test.go
+++ b/formam_test.go
@@ -870,3 +870,28 @@ func TestPanic(t *testing.T) {
 	}
 
 }
+
+// #25
+func TestOverflow(t *testing.T) {
+	s := struct {
+		N uint8
+	}{}
+
+	vals := url.Values{
+		"N": []string{"300"},
+	}
+
+	dec := formam.NewDecoder(&formam.DecoderOptions{})
+	err := dec.Decode(vals, &s)
+	if err == nil {
+		t.Fatalf("error is nil")
+	}
+	if s.N != 0 {
+		t.Fatalf("s.N is %d", s.N)
+	}
+
+	fErr := err.(*formam.Error)
+	if fErr.Code() != formam.ErrCodeRange {
+		t.Fatalf("error code is %d", fErr.Code())
+	}
+}

